Add Logbook Export Functionality to the Search Runs Widget

cailafinn opened the pull request:
**Description of work.**
Adds a button to the Search Runs widget on the Refl gui that allows the user to export the content of the table out to a `.csv` file that can be loaded into a spreadsheet program.

Scientists were using this widget as a logbook rather than an external one, so having something they can save out and take away has been highly requested. 

**To test:**
*NOTE: The implementation of this makes use of the `QFileDialog`, which has [some interesting fights with Jetbrains IDE's](https://stackoverflow.com/questions/57646908/pyqt5-qfiledialog-is-not-returning-correct-paths-in-ubuntu) when launched from there and causing messed up paths. To see a representative example of the code working, launch workbench from a terminal.*

1. Load an experiment (ID: 1120015 Cycle: 11_3 is my usual testing go-to)
2. Type some details into the Comment and exclude columns
3. Hit enter after editing (this bit is important, and a limit of the QTableView, the changes don't get picked up otherwise)
4. Click the export button
5. Think up a name and a place for your logbook
6. Hit enter
7. Check the contents of the CSV is as expected and the correct file extension was applied to the file.  

Fixes #33468 

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Commit:
Generate a CSV from a SearchResults object

RE #33468

Commit:
Handle empty results

RE #33468

Commit:
Refactor to make and get methods

It's nicer to have separation of concerns here. Most of the using
classes just want to use the member variables, but it's nice for testing
to be able to make sure the CSV generation is working with anything
without worrying about other member functions.

RE #33468

Commit:
Add CSV passthrough method to RunsView

RE #33468

Commit:
Allow saving of logbook csvs to file system

RE #33468

Commit:
Add export button to GUI

RE #33468

Commit:
Add handling for .csv missing from filename

Also add error handling for bad paths.

RE #33468

Commit:
Extend tooltip to inform of editing requirements

You have to click out of cells or press enter after editing to make sure
that the updated value is included in the exported logbook. This adds
to the tooltip to inform the user of this.

RE #33468

Commit:
Add release notes

RE #33468

Commit:
Add CPPCheck suppression.

std::accumulate isn't as efficient as a raw loop for concatenating
strings. It makes two copies that arent needeed. This is updated in
C++20, but we aren't there yet.

RE #33468

gemmaguest reviewed (changes requested):
Nice code and some good tests added :+1: :+1:. Just a couple of minor suggestions plus a slight design adjustment to work around Qt not being MVP (it's a lengthy explanation, but should be a simple change!)

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/GUI/Runs/IRunsView.h:
```diff
@@ -80,6 +81,7 @@ class MANTIDQT_ISISREFLECTOMETRY_DLL IRunsView : public MantidQt::MantidWidgets:
   virtual void setSearchResultsColumnWidth(int column, int width) = 0;
   virtual ISearchModel const &searchResults() const = 0;
   virtual ISearchModel &mutableSearchResults() = 0;
+  virtual std::string getLogbookCSV() const = 0;
```
Might be worth having a chat about this anyway in case we can simplify any of it!

Commit:
Remove usage of the term logbook

This is just what the results will likely be used for, but isn't a
guarentee or a specificaiton of the format. There may be an actual
logbook elsewhere.

RE #33468

Commit:
Refactor to fit refl code standards

We prefer to catch errors and return early rather than keeping things in
nested ifs.

RE #33468

Commit:
Avoid passing through view

Use m_searcher instead which is a backdoor into the searching model.

RE #33468

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/GUI/Runs/RunsPresenter.h:
```diff
@@ -16,6 +16,7 @@
 #include "MantidAPI/AlgorithmObserver.h"
 #include "MantidAPI/IAlgorithm.h"
 #include "SearchResult.h"
+#include <fstream>
```
Could this be in the .cpp file?

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/test/Runs/MockRunsView.h:
```diff
@@ -31,6 +31,7 @@ class MockRunsView : public IRunsView {
   MOCK_METHOD2(setSearchResultsColumnWidth, void(int, int));
   MOCK_CONST_METHOD0(searchResults, ISearchModel const &());
   MOCK_METHOD0(mutableSearchResults, ISearchModel &());
+  MOCK_CONST_METHOD0(getSearchResultsCSV, std::string());
```
I think this can be removed now?

gemmaguest reviewed (changes requested):
Thanks for the changes - looks good :+1:. Could we also add some tests for the new functions in QtCatalogSearcher and RunsPresenter? I think the test suites should already be set up with the relevant mocks for the searcher and message handler, so hopefully these are straightforward to add?

ConorMFinn commented:
> Thanks for the changes - looks good 👍. Could we also add some tests for the new functions in QtCatalogSearcher and RunsPresenter? I think the test suites should already be set up with the relevant mocks for the searcher and message handler, so hopefully these are straightforward to add?

I wasn't actually done working on the changes yet, just pushing so I didn't lose anything if my machine died

Commit:
Add tests for export from RunsPresenter

Also does a lot of refactoring so that the RunsPresenter can access the
fileHandler (which is the QtMainWindowView)

RE #33468

Commit:
Add QtCatalogSearcher resultscsv test

RE #33468

gemmaguest reviewed (changes requested):
Code changes look good and some nice tests added :+1:

Just one thing I've noticed during functional testing - if I cancel the file browser dialog, it saves the results anyway to a `.csv` file in the directory I ran mantid from.

Commit:
Stop saving if no filename given

RE #33468

gemmaguest reviewed (approved):
Thanks for the final fix. This all looks good now :+1:. Happy to merge when tests pass :shipit:

ConorMFinn commented:
#33415

gemmaguest commented:
@mantidproject/gatekeepers Could we get this merged? I'd like to get it into nightly for a user to test
